File: gitlab_downloader/migration.py
"""Git migration utilities for author/committer replacement and repo migration."""

from __future__ import annotations

import json
import logging
import os
import subprocess
from collections.abc import Callable
from pathlib import Path

import yaml

from .models import AuthorMapping, CommitterMapping, MigrationConfig

logger = logging.getLogger(__name__)


class MigrationExecutor:
    """Executes git migrations including author/committer replacement."""

    def __init__(self, config: MigrationConfig):
        """Initialize MigrationExecutor with migration config.

        Args:
            config: MigrationConfig instance with migration settings
        """
        self.config = config
        self.source_repos_path = Path(config.source_repos_path)

    def replace_authors(
        self,
        repo_path: str | Path,
        author_mappings: dict[str, AuthorMapping] | None = None,
        progress_callback: Callable[[str], None] | None = None,
    ) -> bool:
        """Replace authors in git history using git filter-branch.

        Args:
            repo_path: Path to the git repository
            author_mappings: Dictionary of author mappings. If None, uses config mappings
            progress_callback: Optional callback for progress updates

        Returns:
            True if replacement succeeded, False otherwise
        """
        if author_mappings is None:
            author_mappings = self.config.author_mappings

        if not author_mappings:
            logger.info("No author mappings provided, skipping author replacement")
            return True

        repo_path = Path(repo_path)
        if not (repo_path / ".git").exists():
            logger.error(f"Not a git repository: {repo_path}")
            return False

        try:
            if progress_callback:
                progress_callback(f"Starting author replacement in {repo_path}")

            # Create environment variables for git filter-branch
            mapping_script = self._create_author_mapping_script(author_mappings)

            # Use git filter-branch to replace authors
            cmd = [
                "git",
                "filter-branch",
                "-f",
                "--env-filter",
                mapping_script,
                "--",
                "--all",
            ]

            result = subprocess.run(
                cmd,
                cwd=str(repo_path),
                capture_output=True,
                text=True,
                timeout=3600,
            )

            if result.returncode != 0:
                logger.error(f"Author replacement failed in {repo_path}: {result.stderr}")
                return False

            if progress_callback:
                progress_callback(f"Author replacement completed for {repo_path}")

            logger.info(f"Author replacement completed for {repo_path}")
            return True

        except subprocess.TimeoutExpired:
            logger.error(f"Author replacement timed out for {repo_path}")
            return False
        except Exception as e:
            logger.error(f"Error during author replacement in {repo_path}: {e}")
            return False

    def replace_committers(
        self,
        repo_path: str | Path,
        committer_mappings: dict[str, CommitterMapping] | None = None,
        progress_callback: Callable[[str], None] | None = None,
    ) -> bool:
        """Replace committers in git history using git filter-branch.

        Args:
            repo_path: Path to the git repository
            committer_mappings: Dictionary of committer mappings. If None, uses config mappings
            progress_callback: Optional callback for progress updates

        Returns:
            True if replacement succeeded, False otherwise
        """
        if committer_mappings is None:
            committer_mappings = self.config.committer_mappings

        if not committer_mappings:
            logger.info("No committer mappings provided, skipping committer replacement")
            return True

        repo_path = Path(repo_path)
        if not (repo_path / ".git").exists():
            logger.error(f"Not a git repository: {repo_path}")
            return False

        try:
            if progress_callback:
                progress_callback(f"Starting committer replacement in {repo_path}")

            # Create environment variables for git filter-branch
            mapping_script = self._create_committer_mapping_script(committer_mappings)

            # Use git filter-branch to replace committers
            cmd = [
                "git",
                "filter-branch",
                "-f",
                "--env-filter",
                mapping_script,
                "--",
                "--all",
            ]

            result = subprocess.run(
                cmd,
                cwd=str(repo_path),
                capture_output=True,
                text=True,
                timeout=3600,
            )

            if result.returncode != 0:
                logger.error(f"Committer replacement failed in {repo_path}: {result.stderr}")
                return False

            if progress_callback:
                progress_callback(f"Committer replacement completed for {repo_path}")

            logger.info(f"Committer replacement completed for {repo_path}")
            return True

        except subprocess.TimeoutExpired:
            logger.error(f"Committer replacement timed out for {repo_path}")
            return False
        except Exception as e:
            logger.error(f"Error during committer replacement in {repo_path}: {e}")
            return False

    def migrate_repository(
        self,
        repo_path: str | Path,
        author_mappings: dict[str, AuthorMapping] | None = None,
        committer_mappings: dict[str, CommitterMapping] | None = None,
        progress_callback: Callable[[str], None] | None = None,
    ) -> bool:
        """Execute complete migration for a repository (authors + committers).

        Args:
            repo_path: Path to the git repository
            author_mappings: Dictionary of author mappings. If None, uses config mappings
            committer_mappings: Dictionary of committer mappings. If None, uses config mappings
            progress_callback: Optional callback for progress updates

        Returns:
            True if migration succeeded, False otherwise
        """
        repo_path = Path(repo_path)

        if not (repo_path / ".git").exists():
            logger.error(f"Not a git repository: {repo_path}")
            return False

        if author_mappings is None:
            author_mappings = self.config.author_mappings
        if committer_mappings is None:
            committer_mappings = self.config.committer_mappings

        if not author_mappings and not committer_mappings:
            logger.info("No mappings provided, skipping migration")
            return True

        if progress_callback:
            progress_callback(f"Starting migration for {repo_path.name}")

        # Combine author and committer scripts into a single filter-branch pass
        # to avoid destroying backup refs with a second -f invocation
        author_script = (
            self._create_author_mapping_script(author_mappings)
            if author_mappings
            else "true"
        )
        committer_script = (
            self._create_committer_mapping_script(committer_mappings)
            if committer_mappings
            else "true"
        )
        combined_script = f"{author_script}\n{committer_script}"

        try:
            cmd = [
                "git",
                "filter-branch",
                "-f",
                "--env-filter",
                combined_script,
                "--",
                "--all",
            ]

            result = subprocess.run(
                cmd,
                cwd=str(repo_path),
                capture_output=True,
                text=True,
                timeout=3600,
            )

            if result.returncode != 0:
                logger.error(f"Migration failed in {repo_path}: {result.stderr}")
                return False

        except subprocess.TimeoutExpired:
            logger.error(f"Migration timed out for {repo_path}")
            return False
        except Exception as e:
            logger.error(f"Error during migration in {repo_path}: {e}")
            return False

        if progress_callback:
            progress_callback(f"Migration completed for {repo_path.name}")

        return True

    @staticmethod
    def _create_mapping_script(
        mappings: dict[str, AuthorMapping] | dict[str, CommitterMapping],
        env_prefix: str,
    ) -> str:
        """Create bash script for git filter-branch author or committer replacement.

        Args:
            mappings: Dictionary of author or committer mappings
            env_prefix: Git env var prefix, either "AUTHOR" or "COMMITTER"

        Returns:
            Bash script as string
        """
        import shlex

        parts = []
        first = True
        for mapping in mappings.values():
            # Properly escape all values to prevent shell injection
            original_name = shlex.quote(mapping.original_name)
            original_email = shlex.quote(mapping.original_email)
            new_name = shlex.quote(mapping.new_name)
            new_email = shlex.quote(mapping.new_email)

            keyword = "if" if first else "elif"
            parts.append(
                f'{keyword} [ "$GIT_{env_prefix}_NAME" = {original_name} ] && '
                f'[ "$GIT_{env_prefix}_EMAIL" = {original_email} ]; then\n'
                f"  export GIT_{env_prefix}_NAME={new_name}\n"
                f"  export GIT_{env_prefix}_EMAIL={new_email}"
            )
            first = False

        if parts:
            script = "\n".join(parts) + "\nfi"
        else:
            script = "true"
        return script

    @staticmethod
    def _create_author_mapping_script(
        mappings: dict[str, AuthorMapping],
    ) -> str:
        """Create bash script for git filter-branch author replacement."""
        return MigrationExecutor._create_mapping_script(mappings, "AUTHOR")

    @staticmethod
    def _create_committer_mapping_script(
        mappings: dict[str, CommitterMapping],
    ) -> str:
        """Create bash script for git filter-branch committer replacement."""
        return MigrationExecutor._create_mapping_script(mappings, "COMMITTER")


class ConfigFileManager:
    """Manages loading and saving migration configuration from/to files."""

    CONFIG_FILENAMES = ["migration_config.json", "migration_config.yaml", "migration_config.yml"]

    @staticmethod
    def load_config(repo_path: str | Path) -> MigrationConfig | None:
        """Load migration config from repository directory.

        Args:
            repo_path: Path to the repository

        Returns:
            MigrationConfig instance if found, None otherwise

        Raises:
            ValueError: If config file is invalid
        """
        repo_path = Path(repo_path)

        for filename in ConfigFileManager.CONFIG_FILENAMES:
            config_file = repo_path / filename
            if not config_file.exists():
                continue

            try:
                with open(config_file, encoding="utf-8") as f:
                    data = yaml.safe_load(f) if filename.endswith(("yaml", "yml")) else json.load(f)

                if not isinstance(data, dict):
                    raise ValueError(f"Config file must be a dictionary, got {type(data).__name__}")

                return ConfigFileManager._validate_and_create_config(data)
            except (json.JSONDecodeError, yaml.YAMLError, ValueError) as e:
                raise ValueError(f"Failed to load config from {config_file}: {e}") from e

        return None

    @staticmethod
    def save_config(
        repo_path: str | Path, config: MigrationConfig, file_format: str = "json"
    ) -> None:
        """Save migration config to repository directory.

        Args:
            repo_path: Path to the repository
            config: MigrationConfig instance to save
            file_format: File format ('json' or 'yaml')

        Raises:
            ValueError: If format is invalid
        """
        if file_format not in ("json", "yaml"):
            raise ValueError(f"Format must be 'json' or 'yaml', got {file_format}")

        repo_path = Path(repo_path)
        filename = f"migration_config.{file_format}"
        config_file = repo_path / filename

        try:
            data = ConfigFileManager._config_to_dict(config)

            if file_format == "json":
                content = json.dumps(data, indent=2)
            else:
                content = yaml.dump(data, default_flow_style=False)

            # Write with restricted permissions (0o600) since config may contain tokens
            fd = os.open(str(config_file), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
            try:
                os.write(fd, content.encode("utf-8"))
            finally:
                os.close(fd)

            logger.info(f"Config saved to {config_file}")
        except Exception as e:
            raise ValueError(f"Failed to save config to {config_file}") from e

    @staticmethod
    def validate_config(data: dict) -> dict:
        """Validate config file schema.

        Args:
            data: Dictionary from config file

        Returns:
            Validated config dictionary

        Raises:
            ValueError: If validation fails
        """
        required_fields = ["source_repos_path", "target_hosting_url", "target_token"]
        for field in required_fields:
            if field not in data:
                raise ValueError(f"Missing required field: {field}")

        # Validate author mappings structure
        if "author_mappings" in data:
            mappings = data["author_mappings"]
            if not isinstance(mappings, dict):
                raise ValueError("author_mappings must be a dictionary")
            for key, mapping in mappings.items():
                if not isinstance(mapping, dict):
                    raise ValueError(f"author_mappings[{key}] must be a dictionary")
                fields = ["original_name", "original_email", "new_name", "new_email"]
                for field in fields:
                    if field not in mapping:
                        raise ValueError(f"author_mappings[{key}] missing field: {field}")

        # Validate committer mappings structure
        if "committer_mappings" in data:
            mappings = data["committer_mappings"]
            if not isinstance(mappings, dict):
                raise ValueError("committer_mappings must be a dictionary")
            for key, mapping in mappings.items():
                if not isinstance(mapping, dict):
                    raise ValueError(f"committer_mappings[{key}] must be a dictionary")
                fields = ["original_name", "original_email", "new_name", "new_email"]
                for field in fields:
                    if field not in mapping:
                        raise ValueError(f"committer_mappings[{key}] missing field: {field}")

        return data

    @staticmethod
    def _validate_and_create_config(data: dict) -> MigrationConfig:
        """Validate config dict and create MigrationConfig instance.

        Args:
            data: Dictionary from config file

        Returns:
            MigrationConfig instance

        Raises:
            ValueError: If validation fails
        """
        validated = ConfigFileManager.validate_config(data)

        # Parse author mappings
        author_mappings: dict[str, AuthorMapping] = {}
        for key, mapping in validated.get("author_mappings", {}).items():
            author_mappings[key] = AuthorMapping(
                original_name=mapping["original_name"],
                original_email=mapping["original_email"],
                new_name=mapping["new_name"],
                new_email=mapping["new_email"],
            )

        # Parse committer mappings
        committer_mappings: dict[str, CommitterMapping] = {}
        for key, mapping in validated.get("committer_mappings", {}).items():
            committer_mappings[key] = CommitterMapping(
                original_name=mapping["original_name"],
                original_email=mapping["original_email"],
                new_name=mapping["new_name"],
                new_email=mapping["new_email"],
            )

        return MigrationConfig(
            source_repos_path=validated["source_repos_path"],
            target_hosting_url=validated["target_hosting_url"],
            target_token=validated["target_token"],
            author_mappings=author_mappings,
            committer_mappings=committer_mappings,
        )

    @staticmethod
    def _config_to_dict(config: MigrationConfig) -> dict:
        """Convert MigrationConfig to dictionary for serialization.

        Args:
            config: MigrationConfig instance

        Returns:
            Dictionary representation of config
        """
        return {
            "source_repos_path": config.source_repos_path,
            "target_hosting_url": config.target_hosting_url,
            "target_token": config.target_token,
            "author_mappings": {
                key: {
                    "original_name": mapping.original_name,
                    "original_email": mapping.original_email,
                    "new_name": mapping.new_name,
                    "new_email": mapping.new_email,
                }
                for key, mapping in config.author_mappings.items()
            },
            "committer_mappings": {
                key: {
                    "original_name": mapping.original_name,
                    "original_email": mapping.original_email,
                    "new_name": mapping.new_name,
                    "new_email": mapping.new_email,
                }
                for key, mapping in config.committer_mappings.items()
            },
        }

File: lib/migration.js
import { access } from 'node:fs/promises';
import { join } from 'node:path';
import { EventEmitter } from 'node:events';
import { runGitCommand } from './cloner.js';

/**
 * @typedef {import('./author-mapper.js').AuthorMapping} AuthorMapping
 * @typedef {import('./author-mapper.js').CommitterMapping} CommitterMapping
 * @typedef {import('./author-mapper.js').MigrationConfig} MigrationConfig
 */

/** Default timeout for filter-branch operations: 1 hour in milliseconds. */
const FILTER_BRANCH_TIMEOUT_MS = 3600 * 1000;

/**
 * Validate that a string does not contain control characters that could
 * break shell scripts (newlines, null bytes, etc.).
 *
 * @param {string} value
 * @param {string} label - Description for error messages
 */
function validateShellSafe(value, label) {
  if (/[\x00-\x1f\x7f]/.test(value)) {
    throw new Error(`${label} contains invalid control characters`);
  }
}

/**
 * Escape a string for use in a single-quoted shell argument.
 * Replaces every single quote with the sequence: '\'' (end quote, escaped quote, start quote).
 *
 * @param {string} s
 * @returns {string} - Shell-safe single-quoted string
 */
function shellQuote(s) {
  return "'" + s.replace(/'/g, "'\\''") + "'";
}

/**
 * Check if a path contains a .git directory (is a git repo).
 *
 * @param {string} repoPath
 * @returns {Promise<boolean>}
 */
async function isGitRepo(repoPath) {
  try {
    await access(join(repoPath, '.git'));
    return true;
  } catch {
    return false;
  }
}

/**
 * Create a bash env-filter script for git filter-branch that rewrites
 * author or committer identity based on mappings.
 *
 * @param {Record<string, AuthorMapping|CommitterMapping>} mappings
 * @param {'AUTHOR'|'COMMITTER'} envPrefix
 * @returns {string} - Bash script fragment
 */
export function createMappingScript(mappings, envPrefix) {
  const entries = Object.values(mappings);
  if (entries.length === 0) return 'true';

  const parts = [];
  let first = true;

  for (const mapping of entries) {
    const values = [
      ['original_name', mapping.original_name],
      ['original_email', mapping.original_email],
      ['new_name', mapping.new_name],
      ['new_email', mapping.new_email],
    ];
    for (const [label, val] of values) {
      if (val) validateShellSafe(val, `Mapping ${label}`);
    }

    const conditions = [];
    if (mapping.original_name) {
      conditions.push(
        `[ "$GIT_${envPrefix}_NAME" = ${shellQuote(mapping.original_name)} ]`
      );
    }
    if (mapping.original_email) {
      conditions.push(
        `[ "$GIT_${envPrefix}_EMAIL" = ${shellQuote(mapping.original_email)} ]`
      );
    }
    if (conditions.length === 0) continue;
    if (!mapping.new_name && !mapping.new_email) continue;

    const keyword = first ? 'if' : 'elif';
    const conditionStr = conditions.join(' && ');
    const newName = shellQuote(mapping.new_name || mapping.original_name);
    const newEmail = shellQuote(mapping.new_email || mapping.original_email);

    parts.push(
      `${keyword} ${conditionStr}; then\n` +
      `  export GIT_${envPrefix}_NAME=${newName}\n` +
      `  export GIT_${envPrefix}_EMAIL=${newEmail}`
    );
    first = false;
  }

  if (parts.length === 0) return 'true';
  return parts.join('\n') + '\nfi';
}

/**
 * Run git filter-branch with the given env-filter script on a repository.
 *
 * @param {string} repoPath - Path to the git repository
 * @param {string} envFilterScript - Bash script for --env-filter
 * @param {object} [options]
 * @param {AbortSignal} [options.signal] - AbortSignal for cancellation
 * @param {number} [options.timeoutMs] - Timeout in ms (default: 1 hour)
 * @returns {Promise<{success: boolean, stdout: string, stderr: string}>}
 */
async function runFilterBranch(repoPath, envFilterScript, options = {}) {
  const { signal, timeoutMs = FILTER_BRANCH_TIMEOUT_MS } = options;

  // Create our own AbortController for timeout if none provided
  const ac = new AbortController();
  const timeoutId = setTimeout(() => ac.abort(new Error('filter-branch timed out')), timeoutMs);

  // If an external signal is provided, propagate its abort
  const onExternalAbort = () => ac.abort(signal.reason);
  if (signal) {
    if (signal.aborted) {
      clearTimeout(timeoutId);
      return { success: false, stdout: '', stderr: 'Aborted' };
    }
    signal.addEventListener('abort', onExternalAbort, { once: true });
  }

  try {
    const args = [
      'git', '-C', repoPath,
      'filter-branch', '-f',
      '--env-filter', envFilterScript,
      '--', '--all',
    ];

    const result = await runGitCommand(args, { signal: ac.signal });
    return {
      success: result.code === 0,
      stdout: result.stdout,
      stderr: result.stderr,
    };
  } catch (err) {
    return {
      success: false,
      stdout: '',
      stderr: err.message || String(err),
    };
  } finally {
    clearTimeout(timeoutId);
    if (signal) {
      signal.removeEventListener('abort', onExternalAbort);
    }
  }
}

// ─── MigrationExecutor ─────────────────────────────────────────────

/**
 * Executes git author/committer rewriting on repositories.
 *
 * Events emitted:
 * - 'progress' (message: string) - progress updates
 * - 'error' (message: string) - error messages
 */
export class MigrationExecutor extends EventEmitter {
  /**
   * @param {MigrationConfig} config
   */
  constructor(config) {
    super();
    this.config = config;
    this.sourcePath = config.source_repos_path;
    // Prevent unhandled 'error' event from crashing the process
    this.on('error', () => {});
  }

  /**
   * Replace authors in git history using filter-branch.
   *
   * @param {string} repoPath
   * @param {Record<string, AuthorMapping>} [authorMappings]
   * @param {object} [options]
   * @param {AbortSignal} [options.signal]
   * @param {(msg: string) => void} [options.onProgress]
   * @returns {Promise<boolean>}
   */
  async replaceAuthors(repoPath, authorMappings, options = {}) {
    const mappings = authorMappings ?? this.config.author_mappings;
    if (!mappings || Object.keys(mappings).length === 0) {
      return true;
    }
    if (!(await isGitRepo(repoPath))) {
      this._emitError(`Not a git repository: ${repoPath}`);
      return false;
    }
    const script = createMappingScript(mappings, 'AUTHOR');
    this._emitProgress(`Replacing authors in ${repoPath}`, options.onProgress);

    const result = await runFilterBranch(repoPath, script, options);
    if (!result.success) {
      this._emitError(`Author replacement failed for ${repoPath}: ${result.stderr}`);
      return false;
    }
    this._emitProgress(`Author replacement complete for ${repoPath}`, options.onProgress);
    return true;
  }

  /**
   * Replace committers in git history using filter-branch.
   *
   * @param {string} repoPath
   * @param {Record<string, CommitterMapping>} [committerMappings]
   * @param {object} [options]
   * @param {AbortSignal} [options.signal]
   * @param {(msg: string) => void} [options.onProgress]
   * @returns {Promise<boolean>}
   */
  async replaceCommitters(repoPath, committerMappings, options = {}) {
    const mappings = committerMappings ?? this.config.committer_mappings;
    if (!mappings || Object.keys(mappings).length === 0) {
      return true;
    }
    if (!(await isGitRepo(repoPath))) {
      this._emitError(`Not a git repository: ${repoPath}`);
      return false;
    }
    const script = createMappingScript(mappings, 'COMMITTER');
    this._emitProgress(`Replacing committers in ${repoPath}`, options.onProgress);

    const result = await runFilterBranch(repoPath, script, options);
    if (!result.success) {
      this._emitError(`Committer replacement failed for ${repoPath}: ${result.stderr}`);
      return false;
    }
    this._emitProgress(`Committer replacement complete for ${repoPath}`, options.onProgress);
    return true;
  }

  /**
   * Execute complete migration: rewrite both authors and committers in a single
   * filter-branch pass to preserve backup refs.
   *
   * @param {string} repoPath
   * @param {object} [options]
   * @param {Record<string, AuthorMapping>} [options.authorMappings]
   * @param {Record<string, CommitterMapping>} [options.committerMappings]
   * @param {AbortSignal} [options.signal]
   * @param {(msg: string) => void} [options.onProgress]
   * @returns {Promise<boolean>}
   */
  async migrateRepository(repoPath, options = {}) {
    const authorMappings = options.authorMappings ?? this.config.author_mappings;
    const committerMappings = options.committerMappings ?? this.config.committer_mappings;

    const hasAuthors = authorMappings && Object.keys(authorMappings).length > 0;
    const hasCommitters = committerMappings && Object.keys(committerMappings).length > 0;

    if (!hasAuthors && !hasCommitters) {
      return true;
    }

    if (!(await isGitRepo(repoPath))) {
      this._emitError(`Not a git repository: ${repoPath}`);
      return false;
    }

    // Build a combined env-filter to run in a single pass
    const authorScript = hasAuthors
      ? createMappingScript(authorMappings, 'AUTHOR')
      : 'true';
    const committerScript = hasCommitters
      ? createMappingScript(committerMappings, 'COMMITTER')
      : 'true';
    const combinedScript = `${authorScript}\n${committerScript}`;

    const repoName = repoPath.split('/').filter(Boolean).pop() || repoPath;
    this._emitProgress(`Migrating repository: ${repoName}`, options.onProgress);

    const result = await runFilterBranch(repoPath, combinedScript, options);
    if (!result.success) {
      this._emitError(`Migration failed for ${repoPath}: ${result.stderr}`);
      return false;
    }
    this._emitProgress(`Migration complete for ${repoName}`, options.onProgress);
    return true;
  }

  /**
   * Emit a progress event and call the optional callback.
   *
   * @param {string} message
   * @param {((msg: string) => void)|undefined} callback
   */
  _emitProgress(message, callback) {
    this.emit('progress', message);
    if (typeof callback === 'function') callback(message);
  }

  /**
   * Emit an error event.
   *
   * @param {string} message
   */
  _emitError(message) {
    this.emit('error', message);
  }
}
